File: couchbase-entity/src/main/java/com/kaufland/generation/RebindMethodGeneration.java
package com.kaufland.generation;

import com.kaufland.model.field.CblFieldHolder;
import com.kaufland.util.TypeUtil;
import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.MethodSpec;

import java.util.List;

public class RebindMethodGeneration {

    public MethodSpec generate(boolean clearMDocChanges) {

        MethodSpec.Builder rebind = MethodSpec.methodBuilder("rebind").
                addParameter(TypeUtil.createMapStringObject(), "doc").
                addStatement("mDoc = new $T()", TypeUtil.createHashMapStringObject()).
                addCode(CblDefaultGeneration.addAddCall("mDoc")).
                addCode(CodeBlock.builder()
                        .beginControlFlow("if(doc != null)")
                        .addStatement("mDoc.putAll(doc)")
                        .endControlFlow().build()).
                addCode(CblConstantGeneration.addAddCall("mDoc"));

        if (clearMDocChanges) {
            rebind.addStatement("mDocChanges = new $T()", TypeUtil.createHashMapStringObject());
        }

        return rebind.build();

    }
}

File: couchbase-entity/src/main/java/com/kaufland/generation/RebindMethodGeneration.kt
package com.kaufland.generation

import com.kaufland.util.TypeUtil
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.FunSpec

class RebindMethodGeneration {

    fun generate(clearMDocChanges: Boolean): FunSpec {

        val explicitType = if (clearMDocChanges) TypeUtil.hashMapStringAny() else TypeUtil.hashMapStringAnyNullable()
        val type = if (clearMDocChanges) TypeUtil.mapStringAny() else TypeUtil.mapStringAnyNullable()
        val rebind = FunSpec.builder("rebind").addParameter( "doc", type).addStatement("mDoc = %T()", explicitType).addCode(CblDefaultGeneration.addAddCall("mDoc")).addCode(CodeBlock.builder()
                .beginControlFlow("if(doc != null)")
                .addStatement("mDoc.putAll(doc)")
                .endControlFlow().build()).addCode(CblConstantGeneration.addAddCall("mDoc"))

        if (clearMDocChanges) {
            rebind.addStatement("mDocChanges = %T()", TypeUtil.hashMapStringAnyNullable())
        }

        return rebind.build()

    }
}
